### PR TITLE
Update Idling to v1alpha2

### PR DESCRIPTION
**Pull request by DirectXMan12**
This updates idling to v1alpha2.  It strips away the idling controller and API types (in favor of https://github.com/openshift/service-idler), and rewrites the proxy logic to use the `Idler` type
instead of annotations (and similarly with `oc idle`).

TODO:
- [x] Remove old controller
- [x] Update proxy
- [x] Update `oc idle` (simple toggle version)
- [x] Update router controller
- [x] Fix e2e tests
- [ ] idling proxy metrics (follow-up PR, if needed)
- [ ] `oc create idler` with previous discovery code (follow-up PR, if needed)
- [ ] `oc adm router` support (follow-up PR)

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@knobunc ptal

**Comment by DirectXMan12** [comment]
@sallyom FYI

File: test/extended/idling/idling.go [hunk 1]
```diff
@@ -246,223 +185,163 @@ var _ = g.Describe("idling and unidling", func() {
...
+
+		err = tryEchoUDP(svc)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		g.By("Waiting until we have endpoints")
+		err = exutil.WaitForEndpointsAvailable(oc, serviceName)
+		o.Expect(err).ToNot(o.HaveOccurred())
 	})
+
+	// TODO: Work out how to make this test work correctly when run on AWS
```
Comment (knobunc) [inline test/extended/idling/idling.go]: Delete this?

File: pkg/cmd/infra/router/template.go [hunk 2]
```diff
@@ -119,6 +127,7 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.Ciphers, "ciphers", util.Env("ROUTER_CIPHERS", ""), "Specifies the cipher suites to use. You can choose a predefined cipher set ('modern', 'intermediate', or 'old') or specify exact cipher suites by passing a : separated list.")
 	flag.BoolVar(&o.StrictSNI, "strict-sni", isTrue(util.Env("ROUTER_STRICT_SNI", "")), "Use strict-sni bind processing (do not use default cert).")
 	flag.StringVar(&o.MetricsType, "metrics-type", util.Env("ROUTER_METRICS_TYPE", ""), "Specifies the type of metrics to gather. Supports 'haproxy'.")
+	flag.BoolVar(&o.EnableUnidling, "enable-unidling", false, "Whether or not to enable support for interacting with the idling API to properly wake up idled services.")
```
Comment (knobunc) [inline pkg/cmd/infra/router/template.go]: Can we make this an env too please?  We'll need to add 'oc adm' support and that's easier with env.

File: pkg/cmd/infra/router/template.go [hunk 3]
```diff
@@ -405,22 +414,52 @@ func (o *TemplateRouterOptions) Run() error {
...
-
-	svcFetcher := templateplugin.NewListWatchServiceLookup(kc.Core(), o.ResyncInterval, o.Namespace)
-	templatePlugin, err := templateplugin.NewTemplatePlugin(pluginCfg, svcFetcher)
+	idlerClient, err := idlerclient.NewForConfig(o.Config.OpenShiftConfig())
 	if err != nil {
 		return err
 	}
 
-	factory := o.RouterSelection.NewFactory(routeclient, projectclient.Project().Projects(), kc)
-	factory.RouteModifierFn = o.RouteUpdate
```
Comment (knobunc) [inline pkg/cmd/infra/router/template.go]: Where did I go?

Comment (DirectXMan12) [reply pkg/cmd/infra/router/template.go]: fixed

**Comment by DirectXMan12** [comment]
seems like the node proxy is now running under the sdn pod, but nobody bound the system:node-proxier role to the openshift-sdn:sdn service account...

**Comment by DirectXMan12** [comment]
specifically: @openshift/sig-networking if the SDN pod is going to run the kube-proxy, it should have the `system:node-proxier` role, presumable.

**Comment by DirectXMan12** [comment]
@smarterclayton if you wanted to comment on the design of the Idler object, please do so.  I'd like to get this merged soon.

**Comment by smarterclayton** [comment]
/hold

@openshift/api-review the API needs review, I plan on looking at it but haven't yet.

File: pkg/cmd/server/apis/config/v1/types.go [hunk 4]
```diff
@@ -1420,6 +1420,21 @@ type ServiceServingCert struct {
...
 
+// UnidlingProxyConfig holds configuration for the unidling network proxy, which triggers unidling
+// on the presence of network traffic.
+type UnidlingProxyConfig struct {
+	// Enabled controls whether or not the unidling proxy is enabled.
+	Enabled bool `json:"enabled"`
+
+	// NFQueueNumber is the number of the NFQueue used to intercept unidling traffic.
+	// It must be unique.
+	NFQueueNumber uint16 `json:"nfQueueNumber"`
```
Comment (smarterclayton) [inline pkg/cmd/server/apis/config/v1/types.go]: These are incredibly specific config values - I'm not sure a human should ever set them.

File: pkg/proxy/unidler/signaler.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+func (s *idlerSignaler) NeedPods(service types.NamespacedName) error {
+	idler, present, err := s.lookup.IdlerByService(service)
+	if err != nil {
+		return err
+	}
+	if !present {
+		return fmt.Errorf("no idler found for service %s", service.String())
+	}
+
+	if idler.Spec.WantIdle == false {
```
Comment (smarterclayton) [inline pkg/proxy/unidler/signaler.go]: I haven't reviewed the API design yet, but unless this becomes a constant you should never be comparing to a boolean, just testing.

Comment (DirectXMan12) [reply pkg/proxy/unidler/signaler.go]: whoops, agreed -- must have been tired when I wrote that ;-)

File: pkg/cmd/server/apis/config/v1/types.go (hunk 4 above)
Comment (DirectXMan12) [reply pkg/cmd/server/apis/config/v1/types.go]: Well, specifically, it needs to never conflict with the two Kubernetes mark bits (one is the masquerade bit, and the other is similiar), and it needs to not conflict with any other rules people set on their systems -- that's why it's configurable.

Comment (smarterclayton) [reply pkg/cmd/server/apis/config/v1/types.go]: Again, this is not something a human should have to set ever.  Can you determine values for this  for 99% of cases that won't conflict?

**Comment by knobunc** [comment]
@smarterclayton is there no way that we can get this for 3.11?  We've been waiting for it so that we can fix the issues that openshift.io has uncovered with the existing unidling (e.g. events getting eaten due to rate limiting).  I had hoped that since @DirectXMan12 had this PR open so early that we would be able to get it for this release.

I understand that the installation is effectively a manual process at the moment, but in 3.12 we would make an operator for it.  Given that this is effectively for internal use, so that we can get some mileage on it, how much work do you need to have done for the installer in 3.11?  And since the API is alpha and subject to change, how thorough does the review need to be?

**Comment by knobunc** [comment]
@deads2k @liggitt -- Is there any chance either of you can review this API so we can get the unidler working for openshift.io?  The current tech preview unidling uses events to pass messages and they get rate limited.  This alpha object would be far better... but it needs API review.

**Comment by deads2k** [comment]
> @deads2k @liggitt -- Is there any chance either of you can review this API so we can get the unidler working for openshift.io? The current tech preview unidling uses events to pass messages and they get rate limited. This alpha object would be far better... but it needs API review.

Looks like it's coming from this repo: https://github.com/openshift/service-idler, but the whole repo has only ever had two pulls.  Was the rest reviewed sometime?  It's pretty unusual not to have an api in https://github.com/openshift/api when we want to use and support it.  I think the only exceptions have been: 1) things originated in origin that we forgot, 2) a toy API we immediately mutated/removed for clusterup.

**Comment by deads2k** [comment]
/hold



> I don't think this is going to make it for 3.11. There's too much unresolved in the install path, the API needs a thorough design review, and most of the reviewers have other critical path items right now.

It looks like the main API controlling the behavior is coming in via  https://github.com/openshift/service-idler .  For APIs that are going to be exposed to end users and that we expect to support (and if I'm reading the API right, we would end up supporting it and at least providing some kind of migration instructions), we expect those to live in https://github.com/openshift/api .  A good test is probably, "is this owned by openshift and does it have a client".

The idea of the API seems reasonable.  A persistent object describing the idled/unidled state of a particular set of resources.  Details about the, particularly around edge case behavior and some potential for conflicting information deserve more discussion.

**Comment by DirectXMan12** [comment]
> Looks like it's coming from this repo: https://github.com/openshift/service-idler, but the whole repo has only ever had two pulls. Was the rest reviewed sometime?

Yes @knobunc has reviewed it, and we've been testing it unofficially for a while now.  Particularly, @sallyom has been using it to rewrite and simplify the OpenShift Online automated idling.

> It's pretty unusual not to have an api in https://github.com/openshift/api when we want to use and support it. I think the only exceptions have been: 1) things originated in origin that we forgot, 2) a toy API we immediately mutated/removed for clusterup.

The CRD approach was originally suggested by @smarterclayton.  I my mind, there's two reasons to keep this as a CRD, at least for the time being.

1. It's easier to iterate -- the path of "iterate as CRD, move to API object when stable" seems pretty reasonable.

2. It's easier to share around as a self-contained building block.  Others upstream have expressed interest in building automated unidling, and having this in a separate repo means we can easily point them at this repo as a place to start.

> The idea of the API seems reasonable. A persistent object describing the idled/unidled state of a particular set of resources. Details about the, particularly around edge case behavior and some potential for conflicting information deserve more discussion.

We have had a lot of that discussion, but I'd be happy to discuss further.  Both when we (@knobunc and I) designed the original system (more on that below), and when we were iterating on the new system, based on what we learned from the existing system.  We know the existing system doesn't work at the scales we need for online (in addition to not handling some corner cases we'd like it to handle), and we'd like to get some mileage on this system in Online before we fully support it (i.e. tech preview for a release or two), but we can't do that until we can deploy both the idling/unidling controller (the easy part -- it's a self-contained pod with a CRD) and an updated unidling proxy with router support (the harder part, since the router and proxy are part of OpenShift proper).

**Comment by deads2k** [comment]
> The CRD approach was originally suggested by @smarterclayton. I my mind, there's two reasons to keep this as a CRD, at least for the time being.
> 
> It's easier to iterate -- the path of "iterate as CRD, move to API object when stable" seems pretty reasonable.
> 
> It's easier to share around as a self-contained building block. Others upstream have expressed interest in building automated unidling, and having this in a separate repo means we can easily point them at this repo as a place to start.

Using a CRD is not a way to skip an API review.  See https://github.com/openshift/api/blob/master/servicecertsigner/v1alpha1/types.go#L82-L88 an example of a CRD that is based on an API reviewed type.

openshift/api has a light footprint and openshift APIs should live there.  Living in openshift/api does not mean that your code has to live in openshift/origin.

**Comment by DirectXMan12** [comment]
As per a discussion with @deads2k, I'm going to file a PR against openshift/api.  The API will continue to be served as a custom resource, but the API definitions will live in `openshift/api` so they can be released together.

**Pull request closed without merging**